# PostgreSQL **Database Setup Guide**

This project focuses mainly on setting up a table on PostgreSQL with the necessary schema/table/user-permissions, with a special focus on recording Boston (MA, USA) crime data.

<br>

### ***SOURCE***
The single dataset used in this project is a much cleaner dataset (provided by [dataquest.io](https://dataquest.io)) and can be used for direct loading into the table.
<br>*This and more (of Boston) can be found at [data.boston.gov](https://data.boston.gov/dataset/crime-incident-reports-august-2015-to-date-source-new-system)*

<br>

### ***GOAL***
<table>
<td>
    <img alt="The Big Picture" src="./data/color_the_big_picture.jpeg" width="400">
</td>
<td>
In this project, I will walk you through:
<br>

1. Creating a PostgreSQL `crimes_db` database
    
2. `crimes` schema (the folder that you see within the database)
    
3. `boston` table with appropriate datatypes.

4. `readonly` group with minimal permissions, and one associated user `dataanalyst`

5. `readwrite` group with minimal permissions, and one associated user `datascientist`.
</td>
</table>

----

Checking current directory to make sure data file is avaiable:

In [1]:
!tree -P *.csv

.
├── data
│   └── boston.csv
└── sql

2 directories, 1 file


<br>
Data file is available in `data` folder.

In [2]:
file_name = !ls data/*.csv
file_name = file_name[0]
file_name

'data/boston.csv'

<br>
Peek into first few lines of the dataset:

In [3]:
!head {file_name} | column -s"," -t

incident_number  offense_code  description                                 date        day_of_the_week  lat          long
1                619           LARCENY ALL OTHERS                          2018-09-02  Sunday           42.35779134  -71.13937053
2                1402          VANDALISM                                   2018-08-21  Tuesday          42.30682138  -71.06030035
3                3410          TOWED MOTOR VEHICLE                         2018-09-03  Monday           42.34658879  -71.07242943
4                3114          INVESTIGATE PROPERTY                        2018-09-03  Monday           42.33418175  -71.07866441
5                3114          INVESTIGATE PROPERTY                        2018-09-03  Monday           42.27536542  -71.09036101
6                3820          M/V ACCIDENT INVOLVING PEDESTRIAN - INJURY  2018-09-03  Monday           42.29019621  -71.07159012
7                724           AUTO THEFT                                  2018-09-03  Monday     

<br>
How many columns does the dataset have?

In [4]:
columns = !head -n 1 {file_name} | tr ',' ' ' | wc -w

columns = int(columns[0])
columns

7

<br>
How many rows of data in the source?

In [5]:
rows = !wc -l {file_name}
int(rows.nlstr.split()[0]) - 1

298329

*298329 rows of crime reports (note that it's 1 less after excluding the header row).*

---
<br>

## **postgres** interaction
I will be interacting with postgres, 2 ways:
* Jupyter cell magic `%sql` with `postgres meta-commands`, mostly for quick checks.
* `psycopg2` python module, for the rest of postgres interactions.

<br><br><br>

### Check if postgres service is running:

In [6]:
!service postgresql status

12/main (port 5432): online


<br>

Alternatively, run `pg_lsclusters` run to list all the postgres clusters!

In [7]:
!pg_lsclusters

Ver Cluster Port Status Owner    Data directory              Log file
12  main    5432 online postgres /var/lib/postgresql/12/main /var/log/postgresql/postgresql-12-main.log


<br>

***Note:*** *If the server is `down`, run:*
```bash
sudo pg_ctlcluster <Ver> <Cluster> start
```
*`<Ver>` and `<Cluster>` information are displayed when you run `pg_lsclusters`.*

<br><br><br>

### Connect to `postgres` via cell magic:

In [8]:
# CONNECTION meta-data
DB_USER = "vdigsdata"  # Existing user in the database (default postgres is the superuser)
DB_PWD = "nicetry!"
DB_HOST = "localhost"  # localhost, if server runs on your machine
DB_PORT = 5432         # Default
DB_NAME = "postgres"   # Default PostgreSQL database

%reload_ext sql
%sql postgresql://{DB_USER}:{DB_PWD}@{DB_HOST}:{DB_PORT}/{DB_NAME}
%config SqlMagic.autocommit=True
%config SqlMagic.displaycon=False
%config SqlMagic.feedback=False

f"Connected to ---> ******:***@{DB_HOST}:{DB_PORT}/{DB_NAME}"

'Connected to ---> ******:***@localhost:5432/postgres'

In [9]:
print("Run a SELECT to make sure just to make sure you're connected:")
%sql SELECT version() "PostgreSQL version";

Run a SELECT to make sure just to make sure you're connected:


[('PostgreSQL 12.5 (Ubuntu 12.5-0ubuntu0.20.04.1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0, 64-bit',)]

<br>

***Existing users:***

In [11]:
# Filer expression ((po)|(data))* looks for users begging with (po) or (data)
%sql \du+ ((po)|(data))*

[('postgres', True, True, True, True, True, -1, None, [], None, True)]

<br><br><br>

## | | CREATE DATABASE

<br>Before creating a new database to my liking, here's a list of existing databases:

*Note that I'm filtering for database names beginning with p/c (`p=postgres (default)`; `c=crimes_db` which isn't created yet!)*

In [14]:
%sql \l (p|c)*

[('postgres', 'postgres', 'UTF8', 'C.UTF-8', 'C.UTF-8', None)]

<br><br>

As per the illustration, I will first create a new database `crimes_db`:

In [16]:
# Common variables in this notebook
DB_NEW = "crimes_db"
DB_SCHEMA = "crimes"
DB_TABLE = "boston"
DB_ROLE_READONLY = "readonly"
DB_ROLE_READWRITE = "readwrite"

In [17]:
%%sql
COMMIT;
DROP DATABASE IF EXISTS {DB_NEW};
COMMIT;

[]

In [18]:
result = %sql CREATE DATABASE {DB_NEW};

*Printing `result` will display `[]` which is synonymous with python's `None`.
<br>It's `None` because it was a clean exit (no errors)
<br><br>`[]` is captured in `result` just to suppress the output.*

In [19]:
result

[]

<br><br>

### Verification:

In [23]:
# psql meta-command \l lists all databases.
# Here, I'm filtering by p/c beginning databases.

%sql \l (p|c)*

[('crimes_db', 'vdigsdata', 'UTF8', 'C.UTF-8', 'C.UTF-8', None),
 ('postgres', 'postgres', 'UTF8', 'C.UTF-8', 'C.UTF-8', None)]

<br><br><br>

## | | CREATE SCHEMA
<br>

***A look at existing schemas:***

In [24]:
# psql meta-command \dn lists all the schemas you have access to.
# Using this is handy over an explicit SELECT schema_name FROM information_schema.schemata;

%sql \dn

[('public', 'postgres')]

<br>

As per the illustration, I need a new schema `crimes` with-in `crimes_db` database.

<br><br><br>
As a first check, this is the current database:

In [30]:
# pgspecial doesn't yet support \conninfo; instead:

%sql SELECT current_database() "Current Database";

[('postgres',)]

<br><br>

***Connect to the newly created database `crimes_db`, then create `crimes` schema to associate it with the new database:***

In [31]:
%sql postgresql://{DB_USER}:{DB_PWD}@{DB_HOST}:{DB_PORT}/{DB_NEW}

result = %sql CREATE SCHEMA IF NOT EXISTS crimes;

*Printing `result` will display `[]` which is synonymous with python's `None`. It's `None` because it was a clean exit (no errors)
<br>`[]` is captured in `result` just to suppress the output.*

In [32]:
result

[]

<br><br>

### Verification:

In [33]:
# Get a list of active connections
connections =  %sql --connections
print("Currently active connections ==>")
for conn in connections:
    print(conn)

# I need this key to close the connection
db_postgres = connections[list(connections)[0]]
# db_postgres.session.close()
# print(f"Closed connection ==> {list(connections)[0]}", end="\n"*2)


Currently active connections ==>
postgresql://vdigsdata:***@localhost:5432/postgres
postgresql://vdigsdata:***@localhost:5432/crimes_db


In [34]:
%sql SELECT current_database() "Current Database";

[('crimes_db',)]

In [35]:
%sql \dn

[('crimes', 'vdigsdata'), ('public', 'postgres')]

Notice the new schema `crimes` is associated with the `Owner` as specified in the connection details?

<br>Alternatively, you can see here that `crimes` schema is indeed associated with `crimes_db` database

In [50]:
%%sql
SELECT schema_name, catalog_name
FROM information_schema.schemata
WHERE schema_name SIMILAR TO '(c|pub)%';

[('public', 'crimes_db'), ('crimes', 'crimes_db')]

<br><br><br>

## | | CREATING TABLES
<br>
Now that the database and the schema are in place, table creation is in order.

<br>What details does table creation need?
* Column names
* Column data types (and possible enumeration types)
* Maximum data type length (both numerical and textual data)

<br><br>
To determine all of the above, I need to assess the dataset in its original encoding.
<br>I will use two bash utilities, `file`, and `encguess` to detect character encoding:

In [51]:
!file --mime-encoding {file_name}

data/boston.csv: utf-8


In [52]:
!encguess {file_name}

data/boston.csv	UTF-8


<br><br><br>Bash's `file` and `encguess` are readily available for quickly detecting a file's encoding.
<br>However, python's `chardet*` module provides character encoding detection:
<br>`chardet*` is a third-party module that you may need to install.

In [53]:
import chardet

with open(file_name, mode='rb') as file:
    raw_bytes = file.read()
    encoding_name =  chardet.detect(raw_bytes)['encoding']
    print(encoding_name)

del chardet  # I don't need this module anymore. Free-up memory!

utf-8


Perfect!
<br>All 3 utilities concur that it is indeed `UTF-8`!

With this information, it should be easier to read files because, by default, python encoding defaults to `UTF-8`.

<br><br><br>

### Identify column names

In [54]:
!head -n 5 {file_name} | column -s"," -t

incident_number  offense_code  description           date        day_of_the_week  lat          long
1                619           LARCENY ALL OTHERS    2018-09-02  Sunday           42.35779134  -71.13937053
2                1402          VANDALISM             2018-08-21  Tuesday          42.30682138  -71.06030035
3                3410          TOWED MOTOR VEHICLE   2018-09-03  Monday           42.34658879  -71.07242943
4                3114          INVESTIGATE PROPERTY  2018-09-03  Monday           42.33418175  -71.07866441


<br>

|Column names in ***bold*** are modified|
|:--|
|incident_number|
|offense_code|
|description|
|date|
|day_of_the_week|
|lat***itude***|
|long***itude***|

<br><br><br>

### Identify data types
Now, I need to dig a little deeper to assess:
* Possible `enumeration data-type's`.
* `Maximum length` of textual data.
* Sufficient length for `numerical` data-types.

<br><br>
A helper function to assist in the process:

In [56]:
import csv

header = []

# Helper function returning unique values of a column
def fetch_distinct_set(filename, idx):
    global header
    with open(filename) as file:
        reader = csv.reader(file)
        header, rows = next(reader), list(reader)
        column_set = set()

        for row in rows:
            column_set.add(row[idx])

        return column_set

# Determine unique values
for column in range(columns):
    column_set = fetch_distinct_set(file_name, column)
    print(f"{header[column]}:    ".rjust(20, ' '),
          f"{len(column_set)} distinct values".ljust(30, ' '),
          f"Longest entry: {(max_cset:=max(column_set))}".ljust(50, ' '),
          f"(of length {len(max_cset)})")

# No need of csv module anymore.
del csv

incident_number:     298329 distinct values         Longest entry: 99999                               (of length 5)
   offense_code:     219 distinct values            Longest entry: 900                                 (of length 3)
    description:     239 distinct values            Longest entry: WEAPON - OTHER - OTHER VIOLATION    (of length 32)
           date:     1177 distinct values           Longest entry: 2018-09-03                          (of length 10)
day_of_the_week:     7 distinct values              Longest entry: Wednesday                           (of length 9)
            lat:     18177 distinct values          Longest entry: 42.39504158                         (of length 11)
           long:     18177 distinct values          Longest entry: -71.17867378                        (of length 12)


In [60]:
print("PostgreSQL INTERGER type (int4) can hold values upto",
      2**32 //2 -1,
      ", \nwhich is good enough for my usecase (total unique incident_number's) now.")

PostgreSQL INTERGER type (int4) can hold values upto 2147483647 , 
which is good enough for my usecase (total unique incident_number's) now.


<br><br><br>

|COLUMN|ANALYSIS|Data-type Identified|
|:--|:--|:--|
|**incident_number**|has unique values for every entry in the dataset, and is a number.|INTEGER PRIMARY KEY (int4)|
|**offense_code**|has 219 distinct numerical codes.<br>These are pre-defined codes and by no means it is exhaustive (as there is always a possibility of defining a new offense code). I will be leaving this as-is instead of an ENUM. `smallint` fits perfect as the maximum value is only 900. Even if the possible values flow into the thousands, it should be fine.|SMALLINT (int2)|
|**description**|Ideally, columns with lesser unique values are good candidates for enumeration types.<br>This column has 239 distinct values out of the total `298,329` data rows, and can save disk space when enumerated.<br>An enumeration that large can actually have a runtime overhead.<br>For now, I will choose to let it be as-is.<br>Current maximum length of this column is only 32.<br>However, to be on a safer side and make room for possible future entries, I will allocate a size of `100` characters.|VARCHAR(100)|
|**date**|This field is of the format `YYYY-MM-DD` (without timestamp)|DATE|
|**day_of_the_week**|is a good candidate for enumeration type as it can contain only 7 unique universally known values.|ENUM|
|**latitude**|`lat`itude values can range anywhere between -90 and +90.<br>The longest entry is `42.39504158` with 2 significant digits (`42`), and an 8-digit precision (`.39504158`).|NUMERIC(10, 8)|
|**longitude**|`long`itude values can range anywhere between -180 and +180.<br>On assessing values for `lat` and `long` (excluding `-` and `.`) of `Boston` alone, `NUMERIC(10,8)` should be a good candidate as the significant digits (digits before decimal point) never exceeds 100, in which case it would be `NUMERIC(11,8)`.<br>*[Take some time to assess the possible values for Boston, google-map way](https://www.google.com/maps/place/Boston,+MA/@42.3273972,-71.1573074,11.5z/data=!4m5!3m4!1s0x89e3652d0d3d311b:0x787cbf240162e8a0!8m2!3d42.3600825!4d-71.0588801)*. The precision (digits after the decimal point) is currently 8 which happens to be 1 more than what google maps currently displays.|NUMERIC(10, 8)|

 <br>

<br><br><br>

### Create Table
Now that the data types and their precision are identified, enumerated type and table creation are in order.
<br>For this, I will create them via python's `psycopg2` module:

In [64]:
# Right now, there are no user-defined data-types
%sql SELECT * FROM pg_catalog.pg_enum;

[]

In [66]:
import psycopg2

conn = psycopg2.connect(f"dbname={DB_NEW} user={DB_USER}")
# conn.autocommit = True
cur = conn.cursor()

query = """SELECT schemaname, tablename, tableowner, tablespace 
             FROM pg_catalog.pg_tables
            LIMIT 4;"""

cur.execute(query)
cur.fetchall()

[('pg_catalog', 'pg_statistic', 'postgres', None),
 ('pg_catalog', 'pg_type', 'postgres', None),
 ('pg_catalog', 'pg_foreign_server', 'postgres', None),
 ('pg_catalog', 'pg_authid', 'postgres', 'pg_global')]

In [67]:
# Create ENUM for day_of_the_week in `crimes` schema
cur.execute("""
    CREATE TYPE crimes.weekday AS ENUM ('Sunday',
                                        'Monday',
                                        'Tuesday',
                                        'Wednesday',
                                        'Thursday',
                                        'Friday',
                                        'Saturday'
    );
    COMMIT;
""")

In [75]:
%sql SELECT * FROM pg_enum;

[(16476, 16475, 1.0, 'Sunday'),
 (16478, 16475, 2.0, 'Monday'),
 (16480, 16475, 3.0, 'Tuesday'),
 (16482, 16475, 4.0, 'Wednesday'),
 (16484, 16475, 5.0, 'Thursday'),
 (16486, 16475, 6.0, 'Friday'),
 (16488, 16475, 7.0, 'Saturday')]

In [68]:
cur.execute("SELECT * FROM pg_enum;")
cur.fetchall()

[(16476, 16475, 1.0, 'Sunday'),
 (16478, 16475, 2.0, 'Monday'),
 (16480, 16475, 3.0, 'Tuesday'),
 (16482, 16475, 4.0, 'Wednesday'),
 (16484, 16475, 5.0, 'Thursday'),
 (16486, 16475, 6.0, 'Friday'),
 (16488, 16475, 7.0, 'Saturday')]

In [76]:
cur.execute("""DROP TABLE IF EXISTS crimes.boston;""")

In [81]:
!ls -l sql/new_table*

-rwxrwxrwx 1 vdigsdata vdigsdata 231 Feb 11 11:03 sql/new_table_boston.sql


In [117]:
!cat ./sql/new_table_boston.sql

CREATE TABLE IF NOT EXISTS crimes.boston (incident_number INTEGER PRIMARY KEY, offense_code SMALLINT, description VARCHAR(100), date DATE, day_of_the_week weekday, latitude NUMERIC(10,8), longitude NUMERIC(10,8));



In [113]:
%sql \! cat ./sql/new_table_boston.sql

[]

In [115]:
%sql commit;

[]

In [116]:
%sql select * from pg_tables where tablename = 'boston';

[]

In [102]:
%sql select * from crimes.boston;

(psycopg2.errors.UndefinedTable) relation "crimes.boston" does not exist
LINE 1: select * from crimes.boston;
                      ^

[SQL: select * from crimes.boston;]
(Background on this error at: http://sqlalche.me/e/13/f405)


In [ ]:
# Create Table
cur.execute("""
    CREATE TABLE IF NOT EXISTS crimes.boston (
        incident_number INTEGER PRIMARY KEY,
        offense_code    SMALLINT,
        description     VARCHAR(100),
        date            DATE,
        day_of_the_week weekday,
        latitude        NUMERIC(10,8),
        longitude       NUMERIC(10,8)
    );
""")

conn.commit()

In [ ]:
cur.execute("SELECT * FROM crimes.boston");
print(cur.description)

In [ ]:
%sql SELECT * FROM crimes.boston;

In [ ]:
%sql SELECT current_user, current_database();

<br><br><br>

### Database and User Permissions
Now that the table is in place, proper access to the table is of utmost importance.

<table>
<td>
    <img alt="This is not good!" src="./data/color_new_users.jpeg" width="400">
</td>
<td>
    <br>A newly created database is accessible to PUBLIC group by default.
    <br>And, a newly created user belongs to the PUBLIC group by default.
    <br>This means, any new user (unless explicit persmissions are granted/revoked) can access a database. This is lead to unintentional consequences.
    <br>This post-it on the left shows PUBLIC groups' access to the new database, by default.
    <br>
    <br>Crime data is sensitive and should be protected.
</td>
</table>


<br>To enable proper security, I would like to follow the `least privilege principle` in creating the necessary user groups and to limit access to the crimes database.

1. *As a first step, I would like to limit `PUBLIC` user groups' access to the database*.
2. *Next, as per the requirement, create a `readonly` user group and allow only read-only access to this database table.*
3. *Create a `readwrite` user group and allow only read-write access to this database table.*
<br><br>
<img alt="This is what I want." src="./data/color_new_users_groups.jpeg" width="400">


In [ ]:
query = f"REVOKE ALL ON DATABASE {DB_NEW} FROM PUBLIC;"

cur.execute(querry)

query = f"GRANT SELECT ON crimes TO {DB_USER_READONLY}"

<br><br><br>

### Create `readonly` user group


In [ ]:
# Finally, close all the connections
db_crimes = connections[list(connections)[0]]
db_crimes = connections[list(connections)[1]]
db_crimes.session.close()

conn.close()

%config SqlMagic.autocommit=False